### PR TITLE
Move code from a header to a .c file

### PR DESCRIPTION
**Pull request by lyakh**
compile- and run-tested on Up^2 with a couple of days old "master" branch. The current "master" doesn't build for me - "8d408eb cmake: add dist target" broke it.

### CODE REVIEW

**Comment by jajanusz** [comment]
> compile- and run-tested on Up^2 with a couple of days old "master" branch. The current "master" doesn't build for me - "8d408eb cmake: add dist target" broke it.

`git clean -fdx`, but please read this first: https://explainshell.com/explain?cmd=git+clean+-fdx

I guess you just have some artifacts from the old buildsystem maybe .tarball-version in repo?

**Comment by lyakh** [comment]
Is this the reason for this build failure:
```
16:17:11,589 INFO  - /localdisk/tools/xtensa/RG-2017.8-linux/XtensaTools/bin/xt-xcc    CMakeFiles/pipeline_new.dir/pipeline_new.c.o CMakeFiles/pipeline_new.dir/pipeline_mocks.c.o CMakeFiles/pipeline_new.dir/pipeline_mocks_rzalloc.c.o CMakeFiles/pipeline_new.dir/__/__/__/__/__/src/audio/pipeline.c.o  -o pipeline_new -T/quickbuild/workspace/24733/SOF_FW/build_ut/test/cmocka/memory_mock.x /localdisk/tools/cmocka/build_X6H3CNL_2017_8/install/lib/libcmocka-static.a 
16:17:11,601 INFO  - /localdisk/tools/xtensa/RG-2017.8-linux/XtensaTools/bin/xt-xcc    CMakeFiles/mixer.dir/mixer_test.c.o CMakeFiles/mixer.dir/mock.c.o CMakeFiles/mixer.dir/comp_mock.c.o CMakeFiles/mixer.dir/__/__/__/__/__/src/audio/buffer.c.o CMakeFiles/mixer.dir/__/__/__/__/__/src/audio/mixer.c.o  -o mixer -T/quickbuild/workspace/24733/SOF_FW/build_ut/test/cmocka/memory_mock.x /localdisk/tools/cmocka/build_X6H3CNL_2017_8/install/lib/libcmocka-static.a -lm 
16:17:11,621 WARN  - CMakeFiles/pipeline_new_allocation.dir/__/__/__/__/__/src/audio/pipeline.c.o: In function `idc_send_msg':
16:17:11,621 WARN  - pipeline.c:(.text+0x353): undefined reference to `arch_idc_send_msg'
16:17:11,628 INFO  - test/cmocka/src/audio/pipeline/CMakeFiles/pipeline_new_allocation.dir/build.make:129: recipe for target 'test/cmocka/src/audio/pipeline/pipeline_new_allocation' failed
16:17:11,628 INFO  - make[2]: Leaving directory '/quickbuild/workspace/24733/SOF_FW/build_ut'
16:17:11,628 WARN  - make[2]: *** [test/cmocka/src/audio/pipeline/pipeline_new_allocation] Error 2
16:17:11,628 INFO  - CMakeFiles/Makefile2:1154: recipe for target 'test/cmocka/src/audio/pipeline/CMakeFiles/pipeline_new_allocation.dir/all' failed
16:17:11,628 WARN  - make[1]: *** [test/cmocka/src/audio/pipeline/CMakeFiles/pipeline_new_allocation.dir/all] Error 2
16:17:11,628 WARN  - make[1]: *** Waiting for unfinished jobs....
```
? This seems to be not a "normal" build, but something cmocka-related. How do I reproduce it locally or why is it failing?

**Comment by jajanusz** [comment]
@lyakh 
https://thesofproject.github.io/latest/developer_guides/unit_tests.html
You need xt-* (xt-xcc/xt-run) toolchain for that, do you have it or only gcc ?

**Comment by lyakh** [comment]
@jajanusz ah, at least now I see what's failing, thanks! I don't have xt-xcc, but it seems relatively easy to "guess" what has to be done to fix this. Only it's a pity it takes so long to get a result... But anyway, I'll probably need a couple of iterations with the CI to get this right, sorry about the electricity bill :-)

**Comment by jajanusz** [comment]
> @jajanusz ah, at least now I see what's failing, thanks! I don't have xt-xcc, but it seems relatively easy to "guess" what has to be done to fix this. Only it's a pity it takes so long to get a result... But anyway, I'll probably need a couple of iterations with the CI to get this right, sorry about the electricity bill :-)

You can try to play with UTs as it may be helpful later, however it may be challenging to fix when you use UTs first time. Feel free to merge that: https://github.com/lyakh/sof/pull/1 +You can feel like a maintainer ;) 

It should automatically update also this PR, I guess it can be the first PR in this repo with multiple commiters, so may server as example for others - for example when one PR depends on another PR

**Comment by lyakh** [comment]
@jajanusz thanks a lot! Looking at the patch it would take me a bit more than a couple of rounds ;-) I thought I would have to create a mock up for the missing symbol.

**Comment by jajanusz** [comment]
> @jajanusz thanks a lot! Looking at the patch it would take me a bit more than a couple of rounds ;-) I thought I would have to create a mock up for the missing symbol.

You can create stubs when dependencies are 1 level, but I guess there are some spagetthi inside so it's easier to strip unused references

**Comment by lyakh** [comment]
@jajanusz sorry, what do you mean "what's changes format?" The patch isn't introducing any functional changes, it only moves code to a .c file to prevent it from being built multiple times. The only thing I can imagine that might be worth adding to the description is *what* code has been moved, i.e. maybe adding an "idc:" prefix, is that what you meant?
CI servers seem to have problems again :-/

**Comment by jajanusz** [comment]
> @jajanusz sorry, what do you mean "what's changes format?" The patch isn't introducing any functional changes, it only moves code to a .c file to prevent it from being built multiple times. The only thing I can imagine that might be worth adding to the description is _what_ code has been moved, i.e. maybe adding an "idc:" prefix, is that what you meant?
> CI servers seem to have problems again :-/

Yes. Something like `idc: move implementation from header to .c file`. I think all our commit use similar format (it's also in kernel)

**Comment by lyakh** [comment]
@jajanusz sure, can do that, but I've just pushed one more similar PR, and that one moves several headers... So there such a prefix would be difficult

**Comment by lyakh** [comment]
@jajanusz There we go: separate commits, with prefixes, one cmacka problem fixed by you, more expected... I'll close the other PR.

**Comment by jajanusz** [comment]
@lyakh It looks like there are some fixes needed for GCC compilers. I'll fix unit tests after you fix compilation for travis/jenkis cis

File: src/include/sof/wait.h [hunk 1]
```diff
@@ -34,48 +34,17 @@
...
 	uint64_t timeout;
 } completion_t;
 
-void arch_wait_for_interrupt(int level);
-
-static inline void wait_for_interrupt(int level)
-{
-	tracev_event(TRACE_CLASS_WAIT, "WFE");
-	wait_atomic_check;
-	arch_wait_for_interrupt(level);
```
Comment (lgirdwood) [inline src/include/sof/wait.h]: please keep the small functions inline, this way we can avoid unnecessary calls and window exceptions on xtensa

Comment (lyakh) [reply src/include/sof/wait.h]: @lgirdwood should be better now

**Comment by lyakh** [comment]
@jajanusz presumably your UT fixes have to go in simultaneously with this PR?

**Comment by jajanusz** [comment]
> @jajanusz presumably your UT fixes have to go in simultaneously with this PR?

Yes. Otherwise QB CI check will fail what is merge-blocker. I'll try to add commit for fixing UTs later today

**Comment by lyakh** [comment]
@jajanusz that would be great, thanks! Otherwise we might have to rebase again, and then again, and... :-)

**Comment by jajanusz** [comment]
> @jajanusz that would be great, thanks! Otherwise we might have to rebase again, and then again, and... :-)

This one was easy: https://github.com/lyakh/sof/pull/2 ;p

**Comment by lgirdwood** [comment]
@lyakh CI scan

**Comment by lyakh** [comment]
> @lyakh CI scan
@lgirdwood Nice, all the issues I've looked at so far are (logically) not mine - I only copied them over from .h to .c. That being said, many of them are real and have to be fixed. Some are also false positives, e.g. an infinite loop in the panic() implementation ;-) What I'm wondering though: hasn't the complete SOF source base been verified by KW? It seems to be a bit strange to me to use it for new submissions without having applied it to the existing code.
Anyway, what would be the proposed way to fix this? The easiest way for me would be to add patches on top - either in a new patch series (would make it even easier to not have to rebase a couple more times, but would possibly hide some bugs), or as additional patches in this series. But the "correct" way would be to first fix bugs, and then move the code... But that would make it complex, because you only see those bugs when you try to move code around. 
I think the best option would be adding patches to this PR on top of the current ones - that way we make sure that we don't miss any bugs. Would that be acceptable?

**Comment by lgirdwood** [comment]
@lyakh yes incremental should be fine.

**Pull request merged**